src/client/driver/command-executors/execute-hybrid-function/node-snapshots.js

```diff
@@ -0,0 +1,116 @@
+import { processScript } from '../../deps/hammerhead';
+import { positionUtils } from '../../deps/testcafe-core';
+import evalFunction from './eval-function';
+
+function sandboxed (fn) {
+    var code = `(${fn.toString()})`;
+
+    code = processScript(code, false);
+
+    return evalFunction(code);
+}
+
+var getAttrs       = sandboxed(element => element.attributes);
+var getChildNodes  = sandboxed(node => node.childNodes);
+var getChildren    = sandboxed(node => node.children);
+var getTextContent = sandboxed(node => node.textContent);
+var getClassName   = sandboxed(element => element.className);
+var getInnerText   = sandboxed(element => element.innerText);
+
+export class NodeSnapshot {
+    constructor (node) {
+        this.nodeType    = node.nodeType;
+        this.textContent = getTextContent(node);
+
+        this.childNodeCount = getChildNodes(node).length;
+        this.hasChildNodes  = !!this.childNodeCount;
+
+        this.childElementCount = NodeSnapshot._getChildElementCount(node);
+        this.hasChildElements  = !!this.childElementCount;
+    }
+
+    static _getChildElementCount (node) {
+        var children = getChildren(node);
+
+        if (children)
+            return children.length;
+
+        // NOTE: doesn't have `children` for non-element nodes =/
+        var childElementCount = 0;
+        var childNodeCount    = node.childNodes.length;
+
+        for (var i = 0; i < childNodeCount; i++) {
+            if (node.childNodes[i].nodeType === 1)
+                childElementCount++;
+        }
+
+        return childElementCount;
+    }
+}
+
+export class ElementSnapshot extends NodeSnapshot {
+    constructor (element) {
+        super(element);
+
+        this.tagName            = element.tagName.toLowerCase();
+        this.visible            = positionUtils.isElementVisible(element);
+        this.focused            = document.activeElement === element;
+        this.attributes         = ElementSnapshot._getAttrsDictionary(element);
+        this.boundingClientRect = ElementSnapshot._getBoundingClientRect(element);
+        this.classNames         = ElementSnapshot._getClassNames(element);
+        this.style              = ElementSnapshot._getStyle(element);
+        this.innerText          = getInnerText(element);
+
+        [
+            'namespaceURI', 'id',
+            'value', 'checked',
+            'scrollWidth', 'scrollHeight', 'scrollLeft', 'scrollTop',
+            'offsetWidth', 'offsetHeight', 'offsetLeft', 'offsetTop',
+            'clientWidth', 'clientHeight', 'clientLeft', 'clientTop'
+        ].forEach(prop => this[prop] = element[prop]);
+    }
+
+    static _getBoundingClientRect (element) {
```

Comment:
Isn't it expected returns object not like DOM method `getBoundingClientRect`? https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIDOMClientRect